### PR TITLE
Transfer Runtime tests to GitHub actions

### PR DESCRIPTION
**Pull request by aanm**
### Description

The current CI setup runs Runtime and Privileged Tests on Jenkins. However, we can achieve the same functionality using GitHub Actions, providing faster feedback and a better user experience. Jenkins, which can sometimes take an hour to allocate nodes just to run the tests, can be replaced with GitHub Actions to obtain quicker results.

### Testing

The jobs are taking less than 15 minutes to complete. The GitHub Actions also offers operations such as re-run a specific focus group which is helpful in case of flakes since it's not necessary to re-run the entire test suite.

- GitHub Action run: https://github.com/cilium/cilium/actions/runs/5003625978

### Notes

- @nbusseneau can we remove `jenkinsfiles/ginkgo-runtime-kernel.Jenkinsfile` or we need to leave it here because of stable branches? If we leave it can we stop triggering it once this PR is merged?

- In the  GitHub Action example run you will see a summary for each Job (for example [here](https://github.com/cilium/cilium/actions/runs/5003625978#summary-13551098106)). This summary is **not** part of the changes on this PR but they will be added in a follow up PR.

- There are some optimizations made specifically for the LVH as part of the GH worfklow. Those optimizations will be removed from this workflow in a follow up PR once they are added as part of the LVH action itself.

### CODE REVIEW

File: test/provision/docker-run-cilium-docker-plugin.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+CILIUM_DOCKER_PLUGIN_IMAGE=${CILIUM_DOCKER_PLUGIN_IMAGE:-cilium/docker-plugin:latest}
+
+set -e
+shopt -s extglob
+
+# Run without sudo if not available (e.g., running as root)
+SUDO=
+if [ ! "$(whoami)" = "root" ] ; then
```
Comment (ldelossa) [inline test/provision/docker-run-cilium-docker-plugin.sh]: nit, `root` is just a name, but `uid=0` is always root, so maybe best to check with `$(id)` or the env variable `$EUID` if you want this script to run in pedantic environments

File: test/provision/docker-run-cilium-docker-plugin.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    SUDO=sudo
+fi
+
+if [ "$1" = "uninstall" ] ; then
+    if [ -n "$(${SUDO} docker ps -a -q -f label=app=cilium-docker)" ]; then
```
Comment (ldelossa) [inline test/provision/docker-run-cilium-docker-plugin.sh]: May want to pull the docker command out into its own string var.

File: test/provision/dns.sh [hunk 3]
```diff
@@ -3,8 +3,8 @@
 # This script update the dns servers to use google ones.
 set -e
 
-sudo systemctl disable systemd-resolved.service
-sudo service systemd-resolved stop
+sudo systemctl disable systemd-resolved.service || true
```
Comment (ldelossa) [inline test/provision/dns.sh:6]: just remove `set -e` instead ?

**Review by ldelossa — APPROVED** [review]
Couple of bash script nits, but looks fine overall.

File: test/provision/dns.sh (hunk 3 above)
Comment (viktor-kurchenko) [reply test/provision/dns.sh:6]: @ldelossa could you please elaborate?

Comment (ldelossa) [reply test/provision/dns.sh:6]: Hmm maybe I'm mistaken, but the '|| true' is there just to handle the case whre the `systemctl` commands fail, but if you just remove `set -e` the script will continue to run, despite the possible failures?

Comment (viktor-kurchenko) [reply test/provision/dns.sh:6]: Gotcha, tnx! I think it makes sense.

File: pkg/mtu/detect_linux.go [hunk 4]
```diff
@@ -16,8 +16,15 @@ import (
 )
 
 const (
-	externalProbeIPv4 = "1.1.1.1"
-	externalProbeIPv6 = "2606:4700:4700::1111"
```
Comment (pchaigno) [inline pkg/mtu/detect_linux.go:20]: How is this related to this PR? Is the commit description leaving some other reason unsaid?

File: .github/workflows/conformance-runtime.yaml [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+      (github.event_name == 'pull_request' && github.event.pull_request.number)
+    }}
+  cancel-in-progress: true
+
+env:
+  check_url: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
+  # List of Runtime tests
+  # f-<ID>-<description>: <Regex that matches the tests in the section>
+  ###
+  f-01-agent: "RuntimeAgent|RuntimeSSHTests"
```
Comment (pchaigno) [inline .github/workflows/conformance-runtime.yaml]: Any reason `f-01-agent` can't just be `agent`?

File: .github/workflows/conformance-runtime.yaml [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+          for filename in *.xml; do cp "${filename}" "../cilium-junits/${{ matrix.focus }}-${filename}"; done;
+
+      - name: Upload JUnits
+        if: ${{ always() }}
+        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # v3.1.2
+        with:
+          name: cilium-junits
+          path: |
+            cilium-junits/*.xml
+          retention-days: 2
```
Comment (pchaigno) [inline .github/workflows/conformance-runtime.yaml:360]: That's not enough to get statistics for http://isogo.to/dashboard-ci AFAIK. What else is needed? Not having those statistics will make the CI Health job a lot harder (as it already is for existing GitHub workflows :disappointed:).

**Review by pchaigno — CHANGES_REQUESTED** [review]
IIUC, we won't have data for http://isogo.to/dashboard-ci for this. I think that's a blocker because it will make the CI Health team's job a lot harder.

File: test/provision/docker-run-cilium-docker-plugin.sh [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+DOCKER_OPTS+=" --label app=cilium-docker"
+
+if [ -n "$(${SUDO} docker ps -a -q -f label=app=cilium-docker)" ]; then
+    echo "Shutting down running Cilium docker-plugin"
+    ${SUDO} docker rm -f cilium-docker || true
+fi
+
+echo "Launching Cilium docker plugin $CILIUM_DOCKER_PLUGIN_IMAGE with params $CILIUM_OPTS"
+${SUDO} docker create --name cilium-docker $DOCKER_OPTS $CILIUM_DOCKER_PLUGIN_IMAGE
```
Comment (qmonnet) [inline test/provision/docker-run-cilium-docker-plugin.sh]: Non-blocking Bash nits:

- `CILIUM_DOCKER_PLUGIN_IMAGE` can be picked up from the env. It's not supposed to contain more than one word anyway, but still it should be between quotes on this line?
- Would be nice to get consistency on the usage of curly braces, as well.
- In general when the shell is bash, the `[[` is recommended over the `[` form for tests.

**Review by qmonnet — APPROVED** [review]
There's the concern raised by Paul, but for what it's worth the rest looks good to me.

File: pkg/mtu/detect_linux.go (hunk 4 above)
Comment (aanm) [reply pkg/mtu/detect_linux.go:20]: Running the unit tests on GH actions were making this flake to happen frequently and I ended up changing as part of the PR to avoid introducing flakes with it.

File: .github/workflows/conformance-runtime.yaml [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+    }}
+  cancel-in-progress: true
+
+env:
+  check_url: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
+  # List of Runtime tests
+  # f-<ID>-<description>: <Regex that matches the tests in the section>
+  ###
+  f-01-agent: "RuntimeAgent|RuntimeSSHTests"
+  # RuntimeAgentChaos Cilium agent Checking for file-descriptor leak
```
Comment (nebril) [inline .github/workflows/conformance-runtime.yaml:26]: Do these comments get regenerated automatically?

**Review by nebril — APPROVED** [review]

File: test/provision/docker-run-cilium.sh [hunk 9]
```diff
@@ -57,3 +57,7 @@ ${SUDO} docker run --name cilium $DOCKER_OPTS $CILIUM_IMAGE /bin/bash -c "groupa
 # Copy Cilium CLI
 ${SUDO} docker cp cilium:/usr/bin/cilium /usr/bin/
 ${SUDO} docker cp cilium:/usr/bin/cilium-bugtool /usr/bin/
+${SUDO} docker cp cilium:/usr/bin/hubble /usr/bin/
+${SUDO} docker cp cilium:/usr/local/bin/clang /usr/bin/
+${SUDO} docker cp cilium:/usr/local/bin/llc /usr/bin/
+${SUDO} docker cp cilium:/usr/local/bin/tc /usr/bin/
```
Comment (joestringer) [inline test/provision/docker-run-cilium.sh]: Heads up that this is likely making some assumption between the container environment vs. github actions with respect to the underlying libraries.. unless we're statically linking clang for instance (I suspect not). Probably fine for a while but may spontaneously break when GitHub goes to roll out a new Ubuntu on GHA workers (probably we're safe for another 1-1.5 years)

File: test/provision/docker-run-cilium-docker-plugin.sh [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+fi
+
+DOCKER_OPTS+=" --label app=cilium-docker"
+
+if [ -n "$(${SUDO} docker ps -a -q -f label=app=cilium-docker)" ]; then
+    echo "Shutting down running Cilium docker-plugin"
+    ${SUDO} docker rm -f cilium-docker || true
+fi
+
+echo "Launching Cilium docker plugin $CILIUM_DOCKER_PLUGIN_IMAGE with params $CILIUM_OPTS"
```
Comment (joestringer) [inline test/provision/docker-run-cilium-docker-plugin.sh]: Is `CILIUM_OPTS` supposed to be `DOCKER_OPTS` here?

File: test/provision/dns.sh (hunk 3 above)
Comment (joestringer) [reply test/provision/dns.sh:6]: Either way works. `-e` is generally best practice. I could see us adding a linter at some point that enforces `-e` on all scripts, since in general that's more likely to catch any sort of unexpected behaviour in bash scripts. Removing `-e` for this specific case just because we know this particular script doesn't need it will make it the exception, which will require some workaround in such a linter. That said, right now we don't have that so :shrug: does it make a difference? Not sure.

File: test/provision/docker-run-cilium-docker-plugin.sh [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    fi
+    exit 0
+fi
+
+DOCKER_OPTS+=" --label app=cilium-docker"
+
+if [ -n "$(${SUDO} docker ps -a -q -f label=app=cilium-docker)" ]; then
+    echo "Shutting down running Cilium docker-plugin"
+    ${SUDO} docker rm -f cilium-docker || true
+fi
```
Comment (joestringer) [inline test/provision/docker-run-cilium-docker-plugin.sh]: This seems duplicate with lines 15-18...? Could it just be unconditionally run before the `uninstall` command check?

File: .github/workflows/conformance-runtime.yaml [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+        uses: cilium/little-vm-helper@0b7d7157dae56a44dc531c852e7756dc671071af # v0.0.3
+        with:
+          provision: 'false'
+          cmd: |
+            cd /host
+            # Install go-junit-report to generate junit files for the
+            # privileged tests.
+            go install github.com/jstemmer/go-junit-report/v2@7fde4641acef5b92f397a8baf8309d1a45d608cc
+            export GOTEST_FORMATTER="/root/go/bin/go-junit-report -set-exit-code -iocopy -out test/privileged-runtime.xml"
+            make tests-privileged NO_COLOR=1
```
Comment (joestringer) [inline .github/workflows/conformance-runtime.yaml]: I think `tparse` or some GHA-friendly formatter would be nicer for developers to handle failures on this testsuite, but this is not a blocker to merge. We can improve on this after getting the base version in the tree.

Comment (joestringer) [reply .github/workflows/conformance-runtime.yaml]: ^ Probably there's some challenge I'm unaware of for getting both the junit formatter and some nice table-based output, but yeah we can investigate+improve on that when we get to it.

File: .github/workflows/conformance-runtime.yaml [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+          --ginkgo.v -- \
+          -cilium.provision=false \
+          --cilium.image=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-ci \
+          -cilium.tag=${{ steps.vars.outputs.sha }}  \
+          --cilium.operator-image=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/operator \
+          -cilium.operator-tag=${{ steps.vars.outputs.sha }} \
+          --cilium.hubble-relay-image=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/hubble-relay-ci \
+          -cilium.hubble-relay-tag=${{ steps.vars.outputs.sha }} \
+          -cilium.operator-suffix=-ci \
+          -cilium.SSHConfig="cat ./cilium-ssh-config.txt"
```
Comment (joestringer) [inline .github/workflows/conformance-runtime.yaml]: nit: Is there a specific pattern to the usage of flags that start with `-` vs. `--` here? They seem inconsistent, but I don't know if this has a functional impact or not.

**Review by joestringer — CHANGES_REQUESTED** [review]
I mostly looked closely at files other than the github workflow and come up with minor nits. I also glanced through the github workflow, mostly assuming the logic is correct and assuming that others would more closely consider any potential security issues with the actions (given your background there, I assume you also considered this :sweat_smile:).

I agree with Paul that we should ensure that this is effective on the CI dashboard, but I don't know exactly how the 2-day retention period interacts with this. Is that enough for CI dashboard to ingest the results so it can point to them? Are we concerned about our ability to reference older CI results and look closer at failure conditions? I'm not exactly sure. But I will say, this seems like it's starting to demonstrate a path to provide GHA failure observability more generically in the dashboard and I very much welcome steps in that direction. If we need to make some practical changes to this PR to improve on the usage workflows from CI dashboard, then by all means let's improve on it. If we need more experience with this workflow in place, then maybe we can merge and improve on top.

The only caveat would be if we merge this and find out that it's unreliable - that might incentivize doing a more staged rollout where we keep the existing jenkins approach for this set of tests for a little while longer until we can transition safely. But if you're confident about the new workflows here then I don't think that needs to be a blocker.

Requested changes just due to some trivial inconsistencies that appear to be in the scripts.

File: test/provision/dns.sh (hunk 3 above)
Comment (aanm) [reply test/provision/dns.sh:6]: I'm fine either case, I'll keep the `|| true` given the linter comment from Joe.

File: test/provision/docker-run-cilium-docker-plugin.sh (hunk 10 above)
Comment (aanm) [reply test/provision/docker-run-cilium-docker-plugin.sh]: yes

File: test/provision/docker-run-cilium.sh (hunk 9 above)
Comment (aanm) [reply test/provision/docker-run-cilium.sh]: Thank you for the heads up, I'll add a comment in the script regarding this

File: .github/workflows/conformance-runtime.yaml (hunk 5 above)
Comment (aanm) [reply .github/workflows/conformance-runtime.yaml]: I thought it would make things simpler but they don't provide any sort of value. I've removed them. Thanks!

File: .github/workflows/conformance-runtime.yaml (hunk 8 above)
Comment (aanm) [reply .github/workflows/conformance-runtime.yaml:26]: No, they were manually generated.

File: .github/workflows/conformance-runtime.yaml (hunk 12 above)
Comment (aanm) [reply .github/workflows/conformance-runtime.yaml]: Absolutely! They will be presented like [this](https://github.com/cilium/cilium/actions/runs/5003625978#summary-13700711221) but the workflow will require one more change, it's not part of this PR. For now it's only exporting the junit output file.

File: .github/workflows/conformance-runtime.yaml (hunk 6 above)
Comment (aanm) [reply .github/workflows/conformance-runtime.yaml:360]: Starting yesterday, the new version of http://isogo.to/dashboard-ci supports reading junits from GH workflows. Once this PR is merged, they will look like this:

image

The `github.com/cilium/cilium/cilium/cmd` lines were read from the junit output.

 I'll write the documentation about this in a follow up PR.

File: .github/workflows/conformance-runtime.yaml (hunk 13 above)
Comment (aanm) [reply .github/workflows/conformance-runtime.yaml]: I honestly thought it matter but it doesn't. I've removed the `--` for the cilium flags.

File: .github/workflows/conformance-runtime.yaml (hunk 6 above)
Comment (pchaigno) [reply .github/workflows/conformance-runtime.yaml:360]: Nice! IIRC the Slack message, this is only for `main` (vs. PR runs) right now, right? It's not ideal, but I think it's enough to alleviate my concerns regarding CI Health debugging.

Comment (aanm) [reply .github/workflows/conformance-runtime.yaml:360]: Correct, only for `main` and stable branches.

Comment (pchaigno) [reply .github/workflows/conformance-runtime.yaml:360]: FYI, looking at the last table from the screenshot, the JUnit looks pretty "broken" (i.e., listing each small step through the test?).

**Review by pchaigno — APPROVED** [review]
My main concern has been addressed. We should have basic statistic collection on running tests. We won't have PR runs but I'm guessing that will come later on (Paul from the future: it did not) and we can do most of the CI Health job with main runs IMO.

**Review by joestringer — APPROVED** [review]

**Review by viktor-kurchenko — APPROVED** [review]

**Comment by aanm** [comment]
@joestringer you brought up a very good point in terms of the github-sec. On the safe side I ended up switching to run the workflow on pull request events instead of `issue_comment`. If we see that it's being unmanageable then we can switch it back to `issue_comment`.

**Comment by aanm** [comment]
/test-1.26-net-next
